logging/logger.go

```diff
@@ -1,116 +1,68 @@
+// Package logging contains log interface and assorted log related utilities
 package logging
 
 import (
 	"fmt"
 	"io"
-	"io/ioutil"
-	"log"
-
-	"github.com/fatih/color"
 
 	"github.com/buildpack/pack/style"
 )
 
-type Logger struct {
-	verbose bool
-	out     *logWriter
-	err     *logWriter
-}
-
-func NewLogger(stdout, stderr io.Writer, verbose, timestamps bool) *Logger {
-	return &Logger{
-		verbose: verbose,
-		out:     newLogWriter(stdout, timestamps),
-		err:     newLogWriter(stderr, timestamps),
-	}
-}
-
-func (l *Logger) printf(w *logWriter, format string, a ...interface{}) {
-	w.Write([]byte(fmt.Sprintf(format+"\n", a...)))
+// Logger defines behavior required by a logging package used by pack libraries
+type Logger interface {
+	Debug(msg string)
+	Debugf(fmt string, v ...interface{})
+	Info(msg string)
+	Infof(fmt string, v ...interface{})
+	Error(msg string)
+	Errorf(fmt string, v ...interface{})
 }
 
-func (l *Logger) Info(format string, a ...interface{}) {
-	l.printf(l.out, format, a...)
+// WithWriter would typically be used to return an io.Writer that can be used to write text that is unadorned
+// with logging information.
+type WithWriter interface {
+	Writer() io.Writer
 }
 
-func (l *Logger) Verbose(format string, a ...interface{}) {
-	if l.verbose {
-		l.printf(l.out, format, a...)
-	}
-}
-
-func (l *Logger) Error(format string, a ...interface{}) {
-	l.printf(l.err, "\n"+style.Error("ERROR: ")+format, a...)
+type LoggerWithWriter interface {
+	Logger
+	WithWriter
 }
 
-func (l *Logger) Tip(format string, a ...interface{}) {
-	l.printf(l.out, style.Tip("Tip: ")+format, a...)
+// Writer will wrap a logging method in an io.Writer
+type Writer struct {
+	logMethod func(string)
+	prefix    string
 }
 
-func (l *Logger) VerboseWriter() *logWriter {
-	if !l.verbose {
-		return nullLogWriter
+// NewWriter takes a log method and creates a Writer. For example
+// w := NewWriter(log.Info)
+func NewWriter(f func(string)) *Writer {
+	return &Writer{
+		logMethod: f,
 	}
-	return l.out
 }
 
-func (l *Logger) RawVerboseWriter() io.Writer {
-	if !l.verbose {
-		return ioutil.Discard
-	}
-	return l.out.rawOut
+// Writes bytes to the embedded log function
+func (w *Writer) Write(buf []byte) (int, error) {
+	w.logMethod(w.prefix + string(buf))
```

Comment:
This prefixes each write instead of prefixing each line. We shouldn't assume that each underlying write corresponds to a log line. Could we allow the logger itself to provide the `io.Writer`?

Reply:
Nevermind. This problem exists in the original implementation, and should be fixed by a different PR.